pkg/virtual/workspaces/registry/rest.go

```diff
@@ -0,0 +1,482 @@
+}
+
+func (s *REST) getInternalNameFromPrettyName(user kuser.Info, prettyName string) (string, error) {
+	list, err := s.crbInformer.Informer().GetIndexer().ByIndex(PrettyNameIndex, prettyName)
+	if err != nil {
+		return "", err
+	}
+	for _, el := range list {
+		if crb, isCRB := el.(*rbacv1.ClusterRoleBinding); isCRB &&
+			len(crb.Subjects) == 1 && crb.Subjects[0].Name == user.GetName() {
+			return crb.Labels[InternalNameLabel], nil
+		}
+	}
+	return "", kerrors.NewNotFound(tenancyv1alpha1.SchemeGroupVersion.WithResource("workspaces").GroupResource(), prettyName)
+}
+
+// List retrieves a list of Workspaces that match label.
+func (s *REST) List(ctx context.Context, options *metainternal.ListOptions) (runtime.Object, error) {
+	user, ok := apirequest.UserFrom(ctx)
+	if !ok {
+		return nil, kerrors.NewForbidden(tenancyv1alpha1.Resource("workspace"), "", fmt.Errorf("unable to list workspaces without a user on the context"))
+	}
+
+	if scope := ctx.Value(WorkspacesScopeKey); scope == personalScope {
+		user = &kuser.DefaultInfo{
+			Name:   user.GetName(),
+			UID:    user.GetUID(),
+			Groups: []string{},
+			Extra:  user.GetExtra(),
+		}
+	}
+
+	labelSelector, _ := InternalListOptionsToSelectors(options)
+	workspaceList, err := s.workspaceLister.List(user, labelSelector)
+	if err != nil {
+		return nil, err
+	}
+
+	if scope := ctx.Value(WorkspacesScopeKey); scope == personalScope {
+		for i, workspace := range workspaceList.Items {
+			var err error
+			workspaceList.Items[i].Name, err = s.getPrettyNameFromInternalName(user, workspace.Name)
+			if err != nil {
+				return nil, err
+			}
+		}
+	}
+
+	return workspaceList, nil
+}
+
+var _ = rest.Getter(&REST{})
+
+// Get retrieves a Workspace by name
+func (s *REST) Get(ctx context.Context, name string, options *metav1.GetOptions) (runtime.Object, error) {
+	opts := metav1.GetOptions{}
+	if options != nil {
+		opts = *options
+	}
+
+	user, ok := apirequest.UserFrom(ctx)
+	if !ok {
+		return nil, kerrors.NewForbidden(tenancyv1alpha1.Resource("workspace"), "", fmt.Errorf("unable to list workspaces without a user on the context"))
+	}
+
+	if scope := ctx.Value(WorkspacesScopeKey); scope == personalScope {
+		internalName, err := s.getInternalNameFromPrettyName(user, name)
+		if err != nil {
+			return nil, err
+		}
+		name = internalName
+	}
+
+	workspace, err := s.workspaceClient.Get(ctx, name, opts)
+	if err != nil {
+		return nil, err
+	}
+
+	if scope := ctx.Value(WorkspacesScopeKey); scope == personalScope {
+		workspace.Name, err = s.getPrettyNameFromInternalName(user, workspace.Name)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	// TODO Filtering by applying the List operation would probably not be necessary anymore
+	// when using a semi-delegated authorizer in the workspaces virtual workspace that would
+	// delegate this authorization to the main KCP instance hosting the workspaces and RBAC rules
```

Comment:
why not check the lister? The lister is filtered, isn't it?

Reply:
Fixed in commit 61481cf05e289ea05d2c6630f78a21271c83789c